components/automate-deployment/pkg/converge/converger.go

```diff
@@ -127,63 +137,482 @@ func StartConverger(opts ...ConvergerOpt) Converger {
 	return converger
 }
 
+func (converger *converger) run() {
+	for {
+		select {
+		case d := <-converger.debug:
+			d.Do(converger)
+		default:
+		}
+
+		select {
+		case <-converger.stop:
+			//TODO(jaym) drain inbox and fail pending requests
+			close(converger.inbox)
+			close(converger.stop)
+			return
+		case req := <-converger.inbox:
+			logrus.Debugf("got message %s while %s", req, converger.currentState)
+			nextState, stopTimer, err := converger.currentState.ProcessMessage(converger, req)
+			if stopTimer {
+				converger.currentState.StopTimer()
+			}
+			converger.setState(nextState)
+			req.SendResponse(err)
+		case d := <-converger.debug:
+			d.Do(converger)
+		case <-converger.currentState.TimeoutChan():
+			logrus.Debugf("timed out while %s", converger.currentState)
+			nextState, _, err := converger.currentState.ProcessMessage(
+				converger,
+				&timeout{})
+			if err != nil {
+				logrus.WithError(err).Error("error processing timeout event")
+			}
+			converger.setState(nextState)
+		}
+	}
+}
+
+func (converger *converger) setState(s state) {
+	if s.String() != converger.currentState.String() {
+		logrus.WithFields(logrus.Fields{
+			"old_state": converger.currentState,
+			"new_state": s,
+		}).Debug("converger state transition")
+	}
+	converger.currentState = s
+}
+
+// state represents the possible states of our converge
+// loop. Currently, we use a small state machine to represent the
+// possible "waiting" states that we might end up in as the result of
+// a converge request.
+type state interface {
+	// ProcessMessage takes the converger and the most recently
+	// received message. The returned state is the next state of
+	// the fsm. The returned error may be send to the sender of
+	// the message. The retruned bool indicates whether we should
+	// reset the current state timeout before moving to the next
+	// state.
+	ProcessMessage(*converger, message) (state, bool, error)
+	// TimeoutChan returns a channel that should delivery a
+	// message when this state has timed out. If a message is
+	// received on the timeout channel a timeout message will be
+	// sent to the current state.
+	TimeoutChan() <-chan time.Time
+	// StopTimer should correctly stop any timers or resources
+	// associated with the TimeoutChan.
+	StopTimer()
+	// String is a convenience method used for logging. states should
+	// have uniq string descriptors so we can more easily
+	// conditionally log.
+	String() string
+}
+
+// message represents the messages that can be sent to our
+// converger. Messages move the converger from one state to another,
+// possibly triggering some system change.
+type message interface {
```

Comment:
I can go either way, I don't think it simplifies ProcessMessage in any fundamental way since you'd still have to have a switch that covers each message type, unless I'm missing something?